plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/join/JoinProcessor.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.ingest.processor.join;
+
+import org.elasticsearch.ingest.Data;
+import org.elasticsearch.ingest.processor.ConfigurationUtils;
+import org.elasticsearch.ingest.processor.Processor;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class JoinProcessor implements Processor {
+
+    public static final String TYPE = "join";
+
+    private final Map<String, String> fields;
+
+    JoinProcessor(Map<String, String> fields) {
+        this.fields = fields;
+    }
+
+    Map<String, String> getFields() {
+        return fields;
+    }
+
+    @Override
+    public void execute(Data data) {
+        for(Map.Entry<String, String> entry : fields.entrySet()) {
+            List<?> list = data.getPropertyValue(entry.getKey(), List.class);
+            if (list == null) {
+                throw new IllegalArgumentException("field [" + entry.getKey() + "] is null, cannot join.");
+            }
+            String joined = list.stream()
```

Comment:
Just a note when back porting this piece of code needs to be changed. I think we should use java8 where possible and in this case back porting is trivial and the isolated. Only in cases where the back port change wouldn't isolated and difficult we should hold back on java8 usage.